pvnet/models/multimodal/multimodal.py

```diff
@@ -56,105 +52,79 @@ def __init__(
+        self.include_sat = sat_encoder is not None
+        self.include_nwp = nwp_encoder is not None
+        self.include_pv = pv_encoder is not None
         self.include_sun = include_sun
         self.embedding_dim = embedding_dim
         self.add_image_embedding_channel = add_image_embedding_channel
 
         super().__init__(history_minutes, forecast_minutes, optimizer, output_quantiles)
 
-        if include_sat:
+        # Number of features expected by the output_network
+        # Add to this as network pices are constructed
+        fusion_input_features = 0
+
+        if self.include_sat:
             # We limit the history to have a delay of 15 mins in satellite data
-            if sat_encoder is None:
-                sat_encoder = image_encoder
-                sat_encoder_kwargs = encoder_kwargs
 
             if sat_history_minutes is None:
                 sat_history_minutes = history_minutes
+
             self.sat_sequence_len = (sat_history_minutes - min_sat_delay_minutes) // 5 + 1
 
             self.sat_encoder = sat_encoder(
                 sequence_length=self.sat_sequence_len,
-                image_size_pixels=sat_image_size_pixels,
-                in_channels=number_sat_channels + add_image_embedding_channel,
-                out_features=encoder_out_features,
-                **sat_encoder_kwargs,
+                in_channels=sat_encoder.keywords["in_channels"] + add_image_embedding_channel,
             )
             if add_image_embedding_channel:
-                self.sat_embed = ImageEmbedding(318, self.sat_sequence_len, sat_image_size_pixels)
+                self.sat_embed = ImageEmbedding(
+                    318, self.sat_sequence_len, self.sat_encoder.image_size_pixels
+                )
+
+            # Update num features
+            fusion_input_features += self.sat_encoder.out_features
 
-        if include_nwp:
+        if self.include_nwp:
             if nwp_history_minutes is None:
                 nwp_history_minutes = history_minutes
             if nwp_forecast_minutes is None:
                 nwp_forecast_minutes = forecast_minutes
             nwp_sequence_len = nwp_history_minutes // 60 + nwp_forecast_minutes // 60 + 1
 
-            self.nwp_encoder = image_encoder(
+            self.nwp_encoder = nwp_encoder(
                 sequence_length=nwp_sequence_len,
-                image_size_pixels=nwp_image_size_pixels,
-                in_channels=number_nwp_channels + add_image_embedding_channel,
-                out_features=encoder_out_features,
-                **encoder_kwargs,
+                in_channels=nwp_encoder.keywords["in_channels"] + add_image_embedding_channel,
             )
             if add_image_embedding_channel:
-                self.nwp_embed = ImageEmbedding(318, nwp_sequence_len, nwp_image_size_pixels)
+                self.nwp_embed = ImageEmbedding(
+                    318, nwp_sequence_len, self.nwp_encoder.image_size_pixels
```

Comment:
Why is the embedding 318? Just curious

Reply:
The nn.Embedding layer used under the hood inside ImageEmbedding expects an ID in the range [0, 1, ...., n-1] when you specify that the number of embeddings is n. The GSP IDs are 1-317 so I set n=318. So the embedding for ID 0 is wasted. Alternatively I could be passing in (GSP ID - 1) instead of (GSP ID).

 I think the 318 solution was inherited from PVNet  v1, and I didn't know the above when I started updating to v2. It would break our current production model to update it now. But I might do so in the future to clean it up